File: Pod/Classes/UIView+LCC.m
//
//  UIView+LCC.m
//  Pods
//
//  Created by Yuki Takahashi on 2015/10/13.
//
//

#import "UIView+LCC.h"
#import "ISMethodSwizzling.h"
#import "NSObject+LCC.h"

@implementation UIView (LCC)

+ (void)load
{
    @autoreleasepool {
        ISSwizzleInstanceMethod([self class],
                                @selector(awakeFromNib),
                                @selector(lcc_awakeFromNib));
    }
}

#pragma mark - Life cycle

- (void)lcc_awakeFromNib
{
    [self lcc_awakeFromNib];
    [self setLcc_isAwakedFromNib:YES];
}

#pragma mark - flgs

- (BOOL)lcc_isAwakedFromNib
{
    return [self lcc_flgForKey:@selector(lcc_isAwakedFromNib)];
}

- (void)setLcc_isAwakedFromNib:(BOOL)lcc_isAwakedFromNib
{
    [self lcc_setFlg:lcc_isAwakedFromNib forKey:@selector(lcc_isAwakedFromNib)];
}

@end

File: Pod/Classes/UIView+LCC.swift
//
//  UIView+LCC.swift
//  Pods
//
//  Created by Yuki Takahashi on 2015/10/13.
//
//

import Foundation

public extension UIView {
    
    public func afterAwakeFromNib(block: () -> ()) {
        try! afterAwakeFromNibSafety(block)
    }
    
    public func afterAwakeFromNibSafety(block: () -> ()) throws {
        if self.lcc_isAwakedFromNib {
            block()
            return
        }
        try self.dynamicType
            .aspect_hookSelector(Selector("awakeFromNib"),
                withOptions: AspectOptions.PositionAfter.union(.OptionAutomaticRemoval),
                usingVoidBlock: { (_) in
                    var key = Selector("lcc_isAwakedFromNib")
                    self.lcc_setFlg(true, forKey: &key)
                    block()
            })
    }
    
}